File: src/main/java/com/searchlab/ingest/Indexer.java
package com.searchlab.ingest;

import com.searchlab.opensearch.IndexBootstrap;
import org.opensearch.client.opensearch.OpenSearchClient;
import org.opensearch.client.opensearch.core.BulkRequest;
import org.opensearch.client.opensearch.core.BulkResponse;
import org.opensearch.client.opensearch.core.bulk.BulkOperation;
import org.opensearch.client.opensearch.core.bulk.IndexOperation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Indexer {

    private static final Logger log = LoggerFactory.getLogger(Indexer.class);

    private final OpenSearchClient client;

    public Indexer(OpenSearchClient client) {
        this.client = client;
    }

    public int index(List<Chunk> chunks, String sourceFilename) throws IOException {
        if (chunks.isEmpty()) {
            return 0;
        }

        List<BulkOperation> operations = new ArrayList<>(chunks.size());
        String ingestedAt = Instant.now().toString();

        for (Chunk chunk : chunks) {
            String chunkId = ChunkId.compute(sourceFilename, chunk.position());

            Map<String, Object> doc = Map.of(
                    "chunk_id",        chunkId,
                    "chunk_text",      chunk.text(),
                    "source_filename", sourceFilename,
                    "page_number",     chunk.pageNumber(),
                    "chunk_position",  chunk.position(),
                    "ingested_at",     ingestedAt
            );

            operations.add(BulkOperation.of(op -> op
                    .index(IndexOperation.of(i -> i
                            .index(IndexBootstrap.INDEX_NAME)
                            .id(chunkId)
                            .document(doc)
                    ))
            ));
        }

        BulkResponse response = client.bulk(BulkRequest.of(r -> r.operations(operations)));

        if (response.errors()) {
            response.items().stream()
                    .filter(item -> item.error() != null)
                    .forEach(item -> log.error("Bulk index error on id={}: {}", item.id(), item.error().reason()));
            throw new IOException("Bulk indexing completed with errors — see logs above");
        }

        log.debug("Indexed {} chunks for '{}'", chunks.size(), sourceFilename);
        return chunks.size();
    }
}

File: service/searchlab/ingest/indexer.py
from datetime import datetime, timezone

from opensearchpy import OpenSearch
from opensearchpy.helpers import bulk

from .chunker import Chunk


def index_chunks(client: OpenSearch, chunks: list[Chunk], source_filename: str, index: str) -> int:
    ingested_at = datetime.now(timezone.utc).isoformat()
    actions = [
        {
            "_index": index,
            "_id": f"{source_filename}_{chunk.position}",
            "_source": {
                "chunk_id": f"{source_filename}_{chunk.position}",
                "chunk_text": chunk.text,
                "source_filename": source_filename,
                "page_number": chunk.page_number,
                "chunk_position": chunk.position,
                "ingested_at": ingested_at,
            },
        }
        for chunk in chunks
    ]
    success, _ = bulk(client, actions)
    return success


def index_corpus_docs(client: OpenSearch, docs: list[dict], index: str) -> int:
    ingested_at = datetime.now(timezone.utc).isoformat()
    actions = [
        {
            "_index": index,
            "_id": doc["_id"],
            "_source": {
                "chunk_id": doc["_id"],
                "chunk_text": doc["title"] + " " + doc["text"],
                "source_filename": doc["_id"],
                "page_number": 0,
                "chunk_position": 0,
                "ingested_at": ingested_at,
            },
        }
        for doc in docs
    ]
    success, _ = bulk(client, actions)
    return success
